crates/wasi-common/src/sys/windows/hostcalls_impl/misc.rs

```diff
@@ -76,12 +167,217 @@ pub(crate) fn clock_time_get(clock_id: wasi::__wasi_clockid_t) -> Result<wasi::_
+    let new_event = make_timeout_event(&timeout_event);
+    events.push(new_event);
+}
+
+fn handle_rw_event(event: FdEventData, out_events: &mut Vec<wasi::__wasi_event_t>) {
+    let size = match event.descriptor {
+        Descriptor::OsHandle(os_handle) => {
+            if event.r#type == wasi::__WASI_EVENTTYPE_FD_READ {
+                os_handle.metadata().map(|m| m.len()).map_err(Into::into)
+            } else {
+                // The spec is unclear what nbytes should actually be for __WASI_EVENTTYPE_FD_WRITE and
+                // the implementation on Unix just returns 0 here, so it's probably fine
+                // to do the same on Windows for now.
+                // cf. https://github.com/WebAssembly/WASI/issues/148
+                Ok(0)
+            }
+        }
+        // We return the only universally correct lower bound, see the comment later in the function.
+        Descriptor::Stdin => Ok(1),
+        // On Unix, ioctl(FIONREAD) will return 0 for stdout/stderr. Emulate the same behavior on Windows.
+        Descriptor::Stdout | Descriptor::Stderr => Ok(0),
+    };
+
+    let new_event = make_rw_event(&event, size);
+    out_events.push(new_event);
+}
+
+fn handle_error_event(
+    event: FdEventData,
+    error: Error,
+    out_events: &mut Vec<wasi::__wasi_event_t>,
+) {
+    let new_event = make_rw_event(&event, Err(error));
+    out_events.push(new_event);
+}
+
 pub(crate) fn poll_oneoff(
     timeout: Option<ClockEventData>,
     fd_events: Vec<FdEventData>,
     events: &mut Vec<wasi::__wasi_event_t>,
-) -> Result<Vec<wasi::__wasi_event_t>> {
-    unimplemented!("poll_oneoff")
+) -> Result<()> {
+    use std::fs::Metadata;
+    use std::thread;
+
+    let timeout = timeout
+        .map(|event| {
+            event
+                .delay
+                .try_into()
+                .map(Duration::from_nanos)
+                .map(|dur| (event, dur))
+        })
+        .transpose()?;
+
+    // With no events to listen, poll_oneoff just becomes a sleep.
+    if fd_events.is_empty() {
+        match timeout {
+            Some((event, dur)) => return Ok(handle_timeout(event, dur, events)),
+            // `poll` invoked with nfds = 0, timeout = -1 appears to be an infinite sleep on Unix
+            // usually meant to be interrupted by a signal. Unfortunately, WASI doesn't currently
+            // support signals and there is no way to interrupt this infinite sleep, so we
+            // intend to return `ENOTSUP`.
+            //
+            // Unfortunately, current implementation of poll_oneoff on Unix relies on
+            // `poll_oneoff` returning `Ok(())` in such case, so we emulate this behavior for now
+            // and will address it in a subsequent PR.
+            None => {
+                // error!("poll_oneoff: invoking with neither timeout nor fds not supported");
+                return Ok(()); // Err(Error::ENOTSUP);
```

Comment:
Given @sunfishcode's merging of the change in WASI, I think we can now return `EINVAL` and file an issue to make the corresponding change to the UNIX implementation.  We can probably simply the above comment to simply say there must be something to wait on for an infinite timeout.

Reply:
Actually, given your change in the other PR to return `EINVAL` from the outer impl, I'd be open to an `unreachable!` here instead too.

Reply:
👍  on change.

Reply:
This has to return an `Ok(())`, as per the comment added in #838. I cleaned up the longish outdated comment from this block.